File: api/app/organizers/interactors/snack/create_snack.rb
module Interactors::Snack
  class CreateSnack < Interactors::BaseInteractor
    include Interactor::ContextValidator

    params :attributes, :user

    validate :attributes, :hash do
      attribute :name, :required
      attribute :icon, :optional
      attribute :description, :optional
      attribute :flavor_ids, :required
    end
    validate :user, :required

    def call
      ActiveRecord::Base.transaction do
        @snack = Snack.create!(attributes.compact.except(:flavor_ids).merge({ user_id: user.id }))
        add_snack_flavors

        context.data = @snack
      end
    end

    private

    def add_snack_flavors
      snack_flavors = attributes[:flavor_ids].uniq.map do |flavor_id|
        {
          snack_id: @snack.id,
          flavor_id: flavor_id
        }
      end

      @snack.snack_flavors.create!(snack_flavors)
    end
  end
end

File: ui/src/graphql/mutations/snack/CreateSnack.ts
import { gql } from 'graphql-tag'

export const CREATE_SNACK_MUTATION = gql`
  mutation CreateSnack($attributes: SnackAttributes!) {
    snack: createSnack(attributes: $attributes) {
      id
      name
      icon
      description
      createdAt
      updatedAt

      snackFlavors {
        id
        flavor {
          id
          name
          icon
        }
      }
    }
  }
`
